File: python/blackroad/tasks.py
"""
BlackRoad Tasks API
"""

from typing import Optional, Dict, Any, List


class TaskAPI:
    """
    Task management API for the 30K agent infrastructure.

    Usage:
        # Dispatch a task
        task = client.tasks.dispatch(
            title="Deploy authentication service",
            description="Deploy Keycloak to Security division",
            priority="urgent",
            division="Security"
        )

        # Get task status
        task = client.tasks.get(task["task_id"])

        # Complete a task
        client.tasks.complete(task["task_id"], result="Deployed successfully")

        # List pending tasks
        pending = client.tasks.list(status="pending")
    """

    def __init__(self, client):
        self._client = client

    def dispatch(
        self,
        title: str,
        description: Optional[str] = None,
        priority: str = "medium",
        division: Optional[str] = None,
        target_level: Optional[int] = None,
        metadata: Optional[Dict] = None
    ) -> Dict[str, Any]:
        """
        Dispatch a new task to the agent hierarchy.

        Args:
            title: Task title
            description: Detailed description
            priority: Priority level (urgent, high, medium, low)
            division: Target division (OS, AI, Security, etc.)
            target_level: Target hierarchy level (2=commanders, 3=managers, 4=workers)
            metadata: Additional task metadata

        Returns:
            Created task dictionary with task_id
        """
        data = {
            "title": title,
            "priority": priority
        }
        if description:
            data["description"] = description
        if division:
            data["division"] = division
        if target_level:
            data["target_level"] = target_level
        if metadata:
            data["metadata"] = metadata

        return self._client.post("/tasks", data=data)

    def get(self, task_id: str) -> Dict[str, Any]:
        """Get a specific task by ID."""
        return self._client.get(f"/tasks/{task_id}")

    def list(
        self,
        status: Optional[str] = None,
        priority: Optional[str] = None,
        division: Optional[str] = None,
        assigned_agent: Optional[str] = None,
        limit: int = 100,
        offset: int = 0
    ) -> List[Dict[str, Any]]:
        """
        List tasks with optional filters.

        Args:
            status: Filter by status (pending, assigned, in_progress, completed, failed)
            priority: Filter by priority
            division: Filter by division
            assigned_agent: Filter by assigned agent
            limit: Maximum results
            offset: Pagination offset

        Returns:
            List of task dictionaries
        """
        params = {"limit": limit, "offset": offset}
        if status:
            params["status"] = status
        if priority:
            params["priority"] = priority
        if division:
            params["division"] = division
        if assigned_agent:
            params["assigned_agent"] = assigned_agent

        return self._client.get("/tasks", params=params).get("tasks", [])

    def complete(self, task_id: str, result: Optional[str] = None) -> Dict[str, Any]:
        """
        Mark a task as completed.

        Args:
            task_id: Task ID
            result: Result summary

        Returns:
            Updated task
        """
        data = {"status": "completed"}
        if result:
            data["result"] = result

        return self._client.put(f"/tasks/{task_id}", data=data)

    def fail(self, task_id: str, reason: Optional[str] = None) -> Dict[str, Any]:
        """
        Mark a task as failed.

        Args:
            task_id: Task ID
            reason: Failure reason

        Returns:
            Updated task
        """
        data = {"status": "failed"}
        if reason:
            data["result"] = reason

        return self._client.put(f"/tasks/{task_id}", data=data)

    def assign(self, task_id: str, agent_id: str) -> Dict[str, Any]:
        """
        Assign a task to a specific agent.

        Args:
            task_id: Task ID
            agent_id: Agent ID to assign

        Returns:
            Updated task
        """
        return self._client.put(f"/tasks/{task_id}", data={
            "assigned_agent": agent_id,
            "status": "assigned"
        })

    def cancel(self, task_id: str) -> Dict[str, Any]:
        """Cancel a pending task."""
        return self._client.delete(f"/tasks/{task_id}")

    def stats(self) -> Dict[str, Any]:
        """Get task statistics."""
        return self._client.get("/tasks/stats")

    def pending(self) -> List[Dict[str, Any]]:
        """Get all pending tasks."""
        return self.list(status="pending")

    def in_progress(self) -> List[Dict[str, Any]]:
        """Get all in-progress tasks."""
        return self.list(status="in_progress")

    def by_division(self, division: str) -> List[Dict[str, Any]]:
        """Get all tasks for a division."""
        return self.list(division=division)

    def urgent(self) -> List[Dict[str, Any]]:
        """Get all urgent tasks."""
        return self.list(priority="urgent")

File: ruby/lib/blackroad/tasks.rb
# frozen_string_literal: true

module Blackroad
  # API for managing tasks
  class TaskAPI
    def initialize(client)
      @client = client
    end

    # Dispatch a new task
    #
    # @param title [String] Task title
    # @param description [String] Task description
    # @param priority [String] Priority (low, medium, high, urgent)
    # @param division [String] Target division
    # @param target_level [Integer] Target agent level
    # @param metadata [Hash] Additional metadata
    # @return [Hash] Created task
    def dispatch(title:, description: nil, priority: "medium", division: nil, target_level: nil, metadata: nil)
      body = { title: title, priority: priority }
      body[:description] = description if description
      body[:division] = division if division
      body[:target_level] = target_level if target_level
      body[:metadata] = metadata if metadata

      @client.post("/tasks", body)
    end

    # Get a specific task by ID
    #
    # @param task_id [String] The task ID
    # @return [Hash] Task data
    def get(task_id)
      @client.get("/tasks/#{task_id}")
    end

    # List tasks with optional filters
    #
    # @param status [String] Filter by status
    # @param priority [String] Filter by priority
    # @param division [String] Filter by division
    # @param limit [Integer] Maximum results
    # @param offset [Integer] Offset for pagination
    # @return [Array<Hash>] List of tasks
    def list(status: nil, priority: nil, division: nil, limit: nil, offset: nil)
      params = {}
      params[:status] = status if status
      params[:priority] = priority if priority
      params[:division] = division if division
      params[:limit] = limit if limit
      params[:offset] = offset if offset

      response = @client.get("/tasks", params)
      response["tasks"] || []
    end

    # Complete a task
    #
    # @param task_id [String] The task ID
    # @param result [String] Completion result/notes
    # @return [Hash] Updated task
    def complete(task_id, result: nil)
      body = { status: "completed" }
      body[:result] = result if result
      @client.put("/tasks/#{task_id}", body)
    end

    # Fail a task
    #
    # @param task_id [String] The task ID
    # @param reason [String] Failure reason
    # @return [Hash] Updated task
    def fail(task_id, reason: nil)
      body = { status: "failed" }
      body[:result] = reason if reason
      @client.put("/tasks/#{task_id}", body)
    end

    # Assign a task to an agent
    #
    # @param task_id [String] The task ID
    # @param agent_id [String] The agent ID
    # @return [Hash] Updated task
    def assign(task_id, agent_id)
      @client.put("/tasks/#{task_id}", { assigned_agent: agent_id, status: "assigned" })
    end

    # Cancel a task
    #
    # @param task_id [String] The task ID
    # @return [Hash] Response
    def cancel(task_id)
      @client.delete("/tasks/#{task_id}")
    end

    # Get task statistics
    #
    # @return [Hash] Statistics
    def stats
      @client.get("/tasks/stats")
    end

    # Get pending tasks
    #
    # @return [Array<Hash>] List of pending tasks
    def pending
      list(status: "pending")
    end

    # Get in-progress tasks
    #
    # @return [Array<Hash>] List of in-progress tasks
    def in_progress
      list(status: "in_progress")
    end

    # Get tasks by division
    #
    # @param division [String] Division name
    # @return [Array<Hash>] List of tasks
    def by_division(division)
      list(division: division)
    end

    # Get urgent tasks
    #
    # @return [Array<Hash>] List of urgent tasks
    def urgent
      list(priority: "urgent")
    end
  end
end
